### ISSUE TITLE
(group) order not preserved

### ISSUE THREAD

**Issue by nschloe** [h5py/h5py#1471]
When inserting data into an h5py file object, the order is not preserved. The items (here: groups) are rather ordered alphabetically:
```python
import h5py

with h5py.File("out.h5", "w") as f:
    a = f.create_group("b")
    b = f.create_group("a")


with h5py.File("out.h5", "r") as f:
    print(f.keys())
```
Output:
```
<KeysViewHDF5 ['a', 'b']>
```
Expected output:
```
<KeysViewHDF5 ['b', 'a']>
```
I don't know if this is a property of HDF5, a problem of the reader. or a problem of the writer.

**Comment by takluyver** [comment]
Preserving order is optional; pass `track_order=True` when creating the file or the group to have it record the order items are created.

https://docs.h5py.org/en/stable/high/file.html#reference

**Comment by nschloe** [comment]
Thanks for the reply. Unfortunately, I your suggestion doesn't work. MWE:
```python
import h5py

print(h5py.get_config().track_order)

with h5py.File("out.h5", "w", track_order=True) as f:
    a = f.create_group("b", track_order=True)
    b = f.create_group("a", track_order=True)


with h5py.File("out.h5", "r", track_order=True) as f:
    print(f.keys())
```
still gives
```
False
<KeysViewHDF5 ['a', 'b']>
```
I believe this issue should be reopened.

**Comment by takluyver** [comment]
So it does. @aparamon, can you take a look? I can reproduce it with 1.10.4.

**Comment by takluyver** [comment]
Does this affect all groups, or only the root group of the file? @1577 may be a duplicate.

**Comment by nschloe** [comment]
Yes, #1577 is a duplicate of this bug. As for subgroups, I fail to create an MWE. Here's what I got, I don't know why `a` doesn't show the subgroups.
```python
import h5py

print(h5py.get_config().track_order)

with h5py.File("out.h5", "w", track_order=True) as f:
    a = f.create_group("b", track_order=True)
    b = f.create_group("a", track_order=True)

    a2 = a.create_group("a2", track_order=True)
    a1 = a.create_group("a1", track_order=True)


with h5py.File("out.h5", "r", track_order=True) as f:
    print(f.keys())
    print(f["a"].keys())
```
```
False
<KeysViewHDF5 ['a', 'b']>
<KeysViewHDF5 []>
```

**Comment by dota17** [comment]
To get subgroups,  you should print like this `print(f["b"].keys())`, because they are inherited from group `b`,

**Comment by nschloe** [comment]
```python
import h5py

print(h5py.get_config().track_order)

with h5py.File("out.h5", "w", track_order=True) as f:
    a = f.create_group("b", track_order=True)
    b = f.create_group("a", track_order=True)

    a2 = a.create_group("a2", track_order=True)
    a1 = a.create_group("a1", track_order=True)


with h5py.File("out.h5", "r", track_order=True) as f:
    print(f.keys())
    print(f["a"].keys())
    print(f["b"].keys())
```
```
False
<KeysViewHDF5 ['a', 'b']>
<KeysViewHDF5 []>
<KeysViewHDF5 ['a2', 'a1']>
```
Indeed it's only in the root level.

**Comment by vasole** [comment]
Hi guys, it works for me if instead of writing f.keys() I write f["/"].keys()

**Comment by dota17** [comment]
yes, work for me too.

**Comment by takluyver** [comment]
Aha - that's starting to make sense. We check if the order was tracked when iterating a group here:

https://github.com/h5py/h5py/blob/75a964bfe3e7ed0ff215435c60a02ebca915f5c3/h5py/h5g.pyx#L118-L121

When you use `f`, `self.grp` will be a low-level FileID object, instead of GroupID. You can set link creation order on a file creation property list, as we do:

https://github.com/h5py/h5py/blob/75a964bfe3e7ed0ff215435c60a02ebca915f5c3/h5py/_hl/files.py#L147-L150

But I guess that internally, when you create the group, this information is moved to the *group* creation property list for the root group. So we probably need to ensure that the code for iterating a group gets a GroupID object one way or another.

**Comment by StoneT2000** [comment]
Has this been fixed yet? It seems that h5py.get_config().track_order() is false still in code despite specifying the root group and file to have track_order=True.

This becomes a problem when I insert data using create_dataset and add "/" to the key name to auto add subgroups e.g.

group.create_dataset("level1/level2/key", data=...)
